readme.md
# Gauge Component for React
A library that provides basic gauge styles and widgets to create gauges efficiently for React projects.

![screen shot](https://raw.githubusercontent.com/JenHsuan/react-gauge-component/master/demo/screen-shot.gif)

## Installation and usage
The easiest way to use react-ui-gauge-component is to install it from npm.

```s
npm i react-ui-gauge-component
```

Then use it in your app

```javascript
import React, { Fragment } from 'react';
import withGauge, { HeartBeat, Needle } from 'react-ui-gauge-component';

const defaultTitleColor = "black";
const defaultValueColor = "black";

const Gauge1 = withGauge({
    WrappedComponent: Needle({value:data[0], min: 0, max: 100}),
    fields: [{title: "Label 1", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor}]
});

const Gauge2 = withGauge({
    WrappedComponent: HeartBeat,
    colorArray: ["#eb001e", "#000", "#3aff20", "#53a5fd", "#aaaaaa"],
    fields: [
        {title: "Label 1", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 2", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor}]
});

const Gauge3 = withGauge({
    WrappedComponent: HeartBeat,
    fields: [
        {title: "Label 1", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 2", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 3", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor}]
});

const Gauge4 = withGauge({
    WrappedComponent: HeartBeat,
    fields: [
        {title: "Label 1", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 2", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 3", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor},
        {title: "Label 4", value: "100", titleColor: defaultTitleColor, valueColor: defaultValueColor}]
});

const DashboardExample = () => {
    return (
        <Fragment>
            <Gauge1/>
            <Gauge2/>
            <Gauge3/>
            <Gauge4/>
        <Fragment/>
    )
}

export default DashboardExample;

```

* [Example for using react-ui-gauge-component with Redux](https://github.com/JenHsuan/example-react-ui-gauge-component/blob/master/components/DashboardExample.js)

## Props
### withGauge
* WrappedComponent -  Put a react component as a widget to compose a new gauge. The default widget is the heartbeat icon.
* colorArray - Put an array of color for the gauge border.

```javascript
colorArray: ["#eb001e", "#f35120", "#3aff20", "#53a5fd", "#aaaaaa"]
```

* fields - Put an array of key, value pairs for the gauge's contents.

```javascript
fields: [
        {title: "Gauge 1", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Gauge 2", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Gauge 3", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Gauge 4", value: "100", titleColor: "black", valueColor: "black"}]
```

### Needle
* value: defult value is zero
* min: defult value is zero
* max: defult value is 100

## License
MIT Licensed. Copyright (c) Jen-Hsuan Hsieh 2021.

## About
* [Author](https://jenhsuan.github.io/ALayman/profile.html)
* [100DaysOfCode](https://dev.to/jenhsuan)
* [Medium](https://medium.com/a-layman)

## Links
* [Change log](https://github.com/JenHsuan/react-gauge-component/blob/master/changelog.md)

## Refernces
* [[Guide] 發佈 npm 套件 - 從手動到自動（0）：專案與套件建置篇](https://pjchender.github.io/2020/02/02/guide-%E7%99%BC%E4%BD%88-npm-%E5%A5%97%E4%BB%B6-%E5%BE%9E%E6%89%8B%E5%8B%95%E5%88%B0%E8%87%AA%E5%8B%95%EF%BC%880%EF%BC%89%EF%BC%9A%E5%B0%88%E6%A1%88%E8%88%87%E5%A5%97%E4%BB%B6%E5%BB%BA%E7%BD%AE%E7%AF%87/)
* [發佈 npm 套件 - 從手動到自動（2）：手動 publish 篇](https://pjchender.blogspot.com/2020/02/guide-npm-2.html)
* [React.js编译](https://chenshenhai.github.io/rollupjs-note/note/chapter04/01.html)
* [如何使用Rollup打包样式文件并添加LiveReload](https://www.w3cplus.com/javascript/learn-rollup-css.html)
* [Babel helpers are bundled into the output file](https://github.com/rollup/rollup-plugin-babel/issues/322)
* [rollup使用过程中遇到的各种问题](https://xudany.github.io/rollup/2019/08/13/rollup%E4%BD%BF%E7%94%A8%E8%BF%87%E7%A8%8B%E4%B8%AD%E9%81%87%E5%88%B0%E7%9A%84%E5%90%84%E7%A7%8D%E9%97%AE%E9%A2%98/)
* [rollup打包js的注意點-haorooms部落格分享](https://www.itread01.com/content/1542163957.html)
* [The complete guide to publishing a React package to npm](https://blog.logrocket.com/the-complete-guide-to-publishing-a-react-package-to-npm/)
* [How does React decide to re-render a component?](https://lucybain.com/blog/2017/react-js-when-to-rerender/)

changelog.md
# Gauge Component for React
## v1.0.0
## V0.0.11: init

webpack.config.js
var path = require('path');
var ExtractTextPlugin = require('extract-text-webpack-plugin');
var webpack = require('webpack');

module.exports = {
  mode: 'development',
  entry: './src/components/SimpleGauge.js',
  output: {
    path: path.resolve(__dirname, 'build'),
    filename: 'SimpleGauge.js',
    library: 'SimpleGauge',
    globalObject: "this",
    libraryTarget: 'commonjs2'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        include: path.join(__dirname, 'src'),
        exclude: path.join(__dirname, '/node_modules/'),
        loader: 'babel-loader',
      }, {
        test: /\.*css$/,
        use : ExtractTextPlugin.extract({
            fallback : 'style-loader',
            use : [
                'css-loader',
                'sass-loader'
            ]
        })
       },
    ]
  },
  plugins: [
    new ExtractTextPlugin("[name].css"),
  ]
};

src/components/SimpleGauge.js
import React, { Fragment } from 'react';

const HeartBeat = () => {
    return (
        <g>
            <path fill="tomato" d="M92.71,7.27L92.71,7.27c-9.71-9.69-25.46-9.69-35.18,0L50,14.79l-7.54-7.52C32.75-2.42,17-2.42,7.29,7.27v0 c-9.71,9.69-9.71,25.41,0,35.1L50,85l42.71-42.63C102.43,32.68,102.43,16.96,92.71,7.27z"></path>
            <animateTransform
                 attributeName="transform"
                 type="scale"
                 values="0.8; 1; 0.9; 1; 1; 0.8;"
                 dur="1s"
                 repeatCount="indefinite">
                </animateTransform>
        </g>
    )
}

const Needle = ({value = 0, min = 0, max = 100}) => () => {
    value = value >= max ? max : value;
    const rotate = Math.floor((value - min) / (max - min) * -180);
    return (
        <g>
            <path className="fmo-gauge__fill"  d="M4.55248,50.95248a47,45,0,1,1,45,45A65,45,0,0,1,4.55248,45.95248Z" fill="none"
                      stroke="#1ccfc9" strokeLinecap="round" strokeMiterlimit="10" strokeWidth="10" strokeDasharray="150,150"/>
            <line fill="none" stroke="#D95857" strokeWidth="4" strokeLinecap="round" strokeMiterlimit="10" x1="50" y1="50" x2="8" y2="50" transform={`translate(50 0) rotate(${rotate} 0 50)`}/>
            <circle fill="#FFFFFF" stroke="#2B2436" strokeWidth="3" strokeMiterlimit="10" cx="50" cy="50" r="5" />
        </g>
    )
}

const withGauge = ({ WrappedComponent = HeartBeat, 
    colorArray = [], 
    fields = [
        {title: "Title 1", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Title 2", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Title 3", value: "100", titleColor: "black", valueColor: "black"},
        {title: "Title 4", value: "100", titleColor: "black", valueColor: "black"}]
    }) => () => {
    const fieldsNumber = fields.length;
    if (fieldsNumber > 4) {
        fields = fields.slice(0 , 4);
    }

    return (
        <svg
            viewBox="0 0 120 120"
            className="chart-gauge"
            fill="none"
            strokeWidth="2">
            <circle 
            strokeDasharray= "200"
            strokeDashoffset= "220"
            stroke = {colorArray && colorArray[0] ? colorArray[0] : "#eb001e"} 
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            <circle 
            strokeDasharray= "200"
            strokeDashoffset= "270"
            stroke = {colorArray && colorArray[1] ? colorArray[1] : "#f35120"} 
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            <circle 
            strokeDasharray= "200"
            strokeDashoffset= "310"
            stroke = {colorArray && colorArray[2] ? colorArray[2] : "#3aff20"} 
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            <circle 
            strokeDasharray= "200"
            strokeDashoffset= "350"
            stroke = {colorArray && colorArray[3] ? colorArray[3] : "#53a5fd"} 
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            <circle 
            strokeDasharray= "200"
            strokeDashoffset= "390"
            stroke = {colorArray && colorArray[4] ? colorArray[4] :"#aaaaaa"} 
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            <circle 
            strokeDasharray = "200"
            strokeDashoffset = "420"
            stroke = "#fff"
            r = "57.32"
            cx = "60"
            cy = "60"
            strokeWidth = "5"/>
            {fieldsNumber === 1 && (
                <Fragment>
                    <text className="gauge-text" x="80" y="35" dx="-20" dy="1em" fill={fields[0].titleColor} fontSize='0.8rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].title}
                    </text>
                    <text className="gauge-text" x="80" y="55" dx="-20" dy="1em" fill={fields[0].valueColor} fontSize='1.1rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].value}
                    </text>
                    <svg width="20" height="20" viewBox="0 0 120 120" x="50" y="90">
                        <WrappedComponent/>
                    </svg>
                </Fragment>
            )}
            {fieldsNumber === 2 && (
                <Fragment>
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="10"
                        y1="60"
                        x2="110"
                        y2="60" />
                        <text className="gauge-text" x="80" y="20" dx="-20" dy="1em" fill={fields[0].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                            {fields[0].title}
                        </text>
                        <text className="gauge-text" x="80" y="38" dx="-20" dy="1em" fill={fields[0].valueColor} fontSize='0.8rem' textAnchor='middle' fontFamily='arial'>
                            {fields[0].value}
                        </text>
                        <text className="gauge-text" x="80" y="70" dx="-20" dy="1em" fill={fields[1].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                            {fields[1].title}
                        </text>
                        <text className="gauge-text" x="80" y="88" dx="-20" dy="1em" fill={fields[1].valueColor} fontSize='0.8rem' textAnchor='middle' fontFamily='arial'>
                            {fields[1].value}
                        </text>
                </Fragment>
            )}
            {fieldsNumber === 3 && (
                <Fragment>
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="10"
                        y1="60"
                        x2="110"
                        y2="60" />
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="60"
                        y1="60"
                        x2="60"
                        y2="110" />
                    <text className="gauge-text" x="80" y="15" dx="-20" dy="1em" fill={fields[0].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].title}
                    </text>
                    <text className="gauge-text" x="80" y="30" dx="-20" dy="1em" fill={fields[0].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].value}
                    </text>
                    <text className="gauge-text" x="57" y="65" dx="-20" dy="1em" fill={fields[1].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[1].title}
                    </text>
                    <text className="gauge-text" x="57" y="82" dx="-20" dy="1em" fill={fields[1].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[1].value}
                    </text>
                    <text className="gauge-text" x="105" y="65" dx="-20" dy="1em" fill={fields[2].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[2].title}
                    </text>
                    <text className="gauge-text" x="105" y="82" dx="-20" dy="1em" fill={fields[2].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[2].value}
                    </text>
                </Fragment>
            )}
            {fieldsNumber === 4 && (
                <Fragment>
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="15"
                        y1="42"
                        x2="105"
                        y2="42" />
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="60"
                        y1="42"
                        x2="60"
                        y2="77" />
                    <line
                        stroke="red"
                        strokeLinecap="round"
                        strokeWidth="1"
                        x1="15"
                        y1="77"
                        x2="105"
                        y2="77" />
                    <text className="gauge-text" x="80" y="10" dx="-20" dy="1em" fill={fields[0].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].title}
                    </text>
                    <text className="gauge-text" x="80" y="20" dx="-20" dy="1em" fill={fields[0].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[0].value}
                    </text>
                    <text className="gauge-text" x="55" y="48" dx="-20" dy="1em" fill={fields[1].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[1].title}
                    </text>
                    <text className="gauge-text" x="55" y="58" dx="-20" dy="1em" fill={fields[1].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[1].value}
                    </text>
                    <text className="gauge-text" x="105" y="48" dx="-20" dy="1em" fill={fields[2].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[2].title}
                    </text>
                    <text className="gauge-text" x="105" y="58" dx="-20" dy="1em" fill={fields[2].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[2].value}
                    </text>
                    <text className="gauge-text" x="80" y="80" dx="-20" dy="1em" fill={fields[3].titleColor} fontSize='0.5rem' textAnchor='middle' fontFamily='arial'>
                        {fields[3].title}
                    </text>
                    <text className="gauge-text" x="80" y="90" dx="-20" dy="1em" fill={fields[3].valueColor} fontSize='0.7rem' textAnchor='middle' fontFamily='arial'>
                        {fields[3].value}
                    </text>
                    <svg width="10" height="10" viewBox="0 0 100 100" x="38" y="93">
                        <WrappedComponent/>
                    </svg>
                </Fragment>
            )}
        </svg>
    )
}

export {
    HeartBeat,
    Needle
};

export default withGauge;
